pkg/lang/ir/types.go

```diff
@@ -50,22 +50,28 @@ type Graph struct {
 
 	VSCodePlugins []vscode.Plugin
 
-	Exec            []string
-	Copy            []CopyInfo
-	Mount           []MountInfo
-	Entrypoint      []string
-	RuntimeCommands map[string]string
-	RuntimeDaemon   [][]string
-	RuntimeEnviron  map[string]string
+	Exec       []string
+	Copy       []CopyInfo
+	Mount      []MountInfo
+	Entrypoint []string
 
 	*JupyterConfig
 	*GitConfig
 	*CondaConfig
 	*RStudioServerConfig
-	*ExposeConfig
 
 	Writer      compileui.Writer
 	CachePrefix string
+
+	RuntimeGraph
+}
+
+// The results during runtime should be maintained here
+type RuntimeGraph struct {
+	RuntimeCommands map[string]string
+	RuntimeDaemon   [][]string
+	RuntimeEnviron  map[string]string
+	RuntimeExpose   *ExposeConfig
```

Comment:
Can we remove the abstraction?
```suggestion
	ExposedPorts []ExposeItem
```

Reply:
> Can we remove the abstraction?

LGTM